File: Backtracking/46_Permutations.cpp
/*
 * @Author: xuezaigds@gmail.com
 * @Last Modified time: 2016-04-19 15:29:06
 */

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        dfs(nums, 0, ans);
        return ans;
    }

    void dfs(vector<int> &num, int begin, vector<vector<int>> &ans){
        if(begin == num.size()){
            ans.push_back(num);
            return;
        }
        for(int i=begin; i<num.size();i++){
            swap(num[begin], num[i]);
            dfs(num, begin+1, ans);
            // Backtrack
            swap(num[begin], num[i]);
        }
    }
};

/*
[]
[1]
[1,2,3]
*/

File: Backtracking/46_Permutations.py
#! /usr/bin/env python
# -*- coding: utf-8 -*-


class Solution(object):
    # Easy to understand: recursively.
    def permute(self, nums):
        ans = []
        self.dfs(nums, [], ans)
        return ans

    def dfs(self, nums, path, ans):
        if not nums:
            ans.append(path)
        for i, n in enumerate(nums):
            self.dfs(nums[:i] + nums[i + 1:], path + [n], ans)


class Solution_2(object):
    # Pythonic way.  recursively.
    # According to: https://leetcode.com/discuss/42550/one-liners-in-python
    def permute(self, nums):
        return [[n] + p
                for i, n in enumerate(nums)
                for p in self.permute(nums[:i] + nums[i + 1:])] or [[]]

"""
[]
[1]
[1,2,3]
"""
